File: android/src/main/kotlin/io/carius/lars/ar_flutter_plugin/CloudAnchorHandler.kt
package io.carius.lars.ar_flutter_plugin

import com.google.ar.core.Anchor
import com.google.ar.core.Anchor.CloudAnchorState
import com.google.ar.core.Session
import java.util.*

// Class for handling logic regarding the Google Cloud Anchor API
internal class CloudAnchorHandler( arSession: Session ) {

    // Listener that can be attached to hosing or resolving processes
    interface CloudAnchorListener {
        // Callback to invoke when cloud anchor task finishes
        fun onCloudTaskComplete(anchorName: String?, anchor: Anchor?)
    }

    private val TAG: String = CloudAnchorHandler::class.java.simpleName
    private val pendingAnchors = HashMap<Anchor, Pair<String?, CloudAnchorListener?>>()
    private val session: Session = arSession

    @Synchronized
    fun hostCloudAnchor(anchorName: String, anchor: Anchor?, listener: CloudAnchorListener?) {
        val newAnchor = session.hostCloudAnchor(anchor)
        // Register listener so it is invoked when the operation finishes
        pendingAnchors[newAnchor] = Pair(anchorName, listener)
    }

    @Synchronized
    fun hostCloudAnchorWithTtl(anchorName: String, anchor: Anchor?, listener: CloudAnchorListener?, ttl: Int) {
        val newAnchor = session.hostCloudAnchorWithTtl(anchor, ttl)
        // Register listener so it is invoked when the operation finishes
        pendingAnchors[newAnchor] = Pair(anchorName, listener)
    }

    @Synchronized
    fun resolveCloudAnchor(anchorId: String?, listener: CloudAnchorListener?) {
        val newAnchor = session.resolveCloudAnchor(anchorId)
        // Register listener so it is invoked when the operation finishes
        pendingAnchors[newAnchor] = Pair(null, listener)
    }

    // Updating function that should be called after each session.update call
    @Synchronized
    fun onUpdate(updatedAnchors: Collection<Anchor>) {
        for (anchor in updatedAnchors) {
            if (pendingAnchors.containsKey(anchor)) {
                if (anchor.cloudAnchorState != CloudAnchorState.NONE && anchor.cloudAnchorState != CloudAnchorState.TASK_IN_PROGRESS){
                    val element: Pair<String?, CloudAnchorListener?>? = pendingAnchors.remove(anchor)
                    element!!.second!!.onCloudTaskComplete(element.first, anchor)
                }
            }
        }
    }

    // Remove all listeners
    @Synchronized
    fun clearListeners() {
        pendingAnchors.clear()
    }

}

File: ios/Classes/CloudAnchorHandler.swift
import Foundation
import ARCoreCloudAnchors

// Listener that can be attached to hosing or resolving processes
protocol CloudAnchorListener {
    // Callback to invoke when cloud anchor task finishes
    func onCloudTaskComplete(anchorName: String?, anchor: GARAnchor?)
}

// Class for handling logic regarding the Google Cloud Anchor API
class CloudAnchorHandler: NSObject, GARSessionDelegate {
    private var session: GARSession
    private var pendingAnchors = [GARAnchor: (String?, CloudAnchorListener?)]()
    
    init(session: GARSession){
        self.session = session
    }
    
    func hostCloudAnchor(anchorName: String, anchor: ARAnchor, listener: CloudAnchorListener?) {
        do {
            let newAnchor = try self.session.hostCloudAnchor(anchor)
            // Register listener so it is invoked when the operation finishes
            pendingAnchors[newAnchor] = (anchorName, listener)
        } catch {
            print(error)
        }
    }
    
    func hostCloudAnchorWithTtl(anchorName: String, anchor: ARAnchor, listener: CloudAnchorListener?, ttl: Int) {
        do {
            let newAnchor = try self.session.hostCloudAnchor(anchor, ttlDays: ttl)
            // Register listener so it is invoked when the operation finishes
            pendingAnchors[newAnchor] = (anchorName, listener)
        } catch {
            print(error)
        }
    }
    
    func resolveCloudAnchor(anchorId: String, listener: CloudAnchorListener?) {
        do {
            let newAnchor = try self.session.resolveCloudAnchor(anchorId)
            // Register listener so it is invoked when the operation finishes
            pendingAnchors[newAnchor] = (nil, listener)
        } catch {
            print(error)
        }
        
    }
        
    func session(_ session: GARSession, didHost anchor: GARAnchor) {
        pendingAnchors[anchor]?.1?.onCloudTaskComplete(anchorName: pendingAnchors[anchor]?.0, anchor: anchor)
    }
    
    func session(_ session: GARSession, didFailToHost anchor: GARAnchor) {
        pendingAnchors[anchor]?.1?.onCloudTaskComplete(anchorName: pendingAnchors[anchor]?.0, anchor: anchor)
    }
    
    func session(_ session: GARSession, didResolve anchor: GARAnchor) {
        pendingAnchors[anchor]?.1?.onCloudTaskComplete(anchorName: pendingAnchors[anchor]?.0, anchor: anchor)
    }
    
    func session(_ session: GARSession, didFailToResolve anchor: GARAnchor) {
        pendingAnchors[anchor]?.1?.onCloudTaskComplete(anchorName: pendingAnchors[anchor]?.0, anchor: anchor)
    }
    
    // Remove all listeners
    func clearListeners() {
        pendingAnchors.removeAll()
    }
}
